Submit transactions with Tickets (RIPD-837):

scottschurr opened the pull request:
There are a number of changes to Tickets with this commit:

 o A Ticket can be used in place of a Sequence in a transaction.

 o The external JSON representation of a Ticket has changed from
   a 256-bit hash to a two-entry object.

 o In TicketCreate and TicketCancel transactions it is not valid
   to use a Ticket instead of a Sequence.

 o Transactions with Tickets sort behind transactions without
   Tickets when they are applied to the ledger.  This will tend
   to keep TicketCreate and TicketCancel operations in front of
   transactions that consume Tickets.

 o There are C++ unit tests for Ticket functionality.

Work remaining:

 o There is some question about whether Tickets need to be
   threaded.  If they do, then threading will need to be added.

 o Documentation needs to be amended.

 o There may be additional fallout when David reviews the code.

vinniefalco commented:
Names of programmers and "meta discussion" should not appear in commit comments (e.g. Work Remaining section)

scottschurr commented:
Reviewers: @JoelKatz for security, @nbougalis for Tickets, and @vinniefalco for general code niceness.

This pull request affects transaction processing, since it changes the order of some transaction verification checks.

vinniefalco commented:
Why is transaction processing affected if tickets are disabled?

scottschurr commented:
Transaction processing is affected because the order of validation checks have changed in `Transactor::checkSeq()`: https://github.com/ripple/rippled/pull/1131/files#diff-63e38aef10c237031210901286f42aa2R162

vinniefalco commented:
@scottschurr Can we arrange this code so that if tickets are disabled, tx processing is not changed?

scottschurr commented:
@vinniefalco I rearranged the checks so we check for `tef` errors, which will stop the transaction, before we check for `tec` errors.  I assumed that if a `tef` error was discovered, then that would be considered a more important failure than a `tec`.

vinniefalco commented on src/ripple/app/tx/impl/LocalTxs.cpp:
```diff
@@ -61,48 +61,83 @@ class LocalTx
     RippleAddress const& getAccount () const
     {
-        return m_account;
+        return account_;
```
This function should return `AccountID`

vinniefalco commented on src/ripple/app/tx/impl/LocalTxs.cpp:
```diff
@@ -61,48 +61,83 @@ class LocalTx
+    STTx::pointer txn_;
+    LedgerIndex   expire_;
+    uint256       id_;
+    RippleAddress account_;
```
The type should be `AccountID`

nbougalis commented on src/ripple/app/misc/CanonicalTXSet.h:
```diff
@@ -39,36 +39,54 @@ class CanonicalTXSet
+
+        inline friend bool operator>  (Key const& lhs, Key const& rhs)
+        {
+            return (operator< (rhs, lhs));
```
Not sure I like this explicit invocation. Why not simply: `return rhs < lhs`?

scottschurr replied:
Good call.  Thanks.  I'll make that change.

nbougalis commented on src/ripple/app/tx/impl/CancelTicket.cpp:
```diff
@@ -41,49 +42,23 @@ class CancelTicket
+        // A Ticket may not be used to cancel a Ticket.  So a Sequence of
+        // zero (which indicates that a Ticket is being used) is malformed.
+        if (mTxn.getSequence () == 0)
+            return tefPAST_SEQ;
```
Returning `tefPAST_SEQ` is incorrect. a `tef` code should be returned when a transaction can't be applied because of the current state of the ledger but that could be applied under some imagined ledger.

I suggest `temBAD_SEQUENCE` instead.

sublimator replied:
>> theoretically, be applied under some other, imagined ledger
In general I thought tef* were of a final disposition ?  

https://wiki.ripple.com/index.php?title=Transaction_errors&oldid=8702#Failure

sublimator replied:
Oh I just saw Scott's comment in an above commit:

```
Historically if a Transaction Sequence of zero (0) was passed,
rippled returned tefPAST_SEQ.  This struck me as not quite right.
Under no circumstance is a Transaction Sequence of zero ever valid
without a Ticket.  So, in reality, the transaction did not fail
(could succeed in an imagined ledger).  The transaction is malformed
if it has a Transaction Sequence of zero (0) and no Ticket -- the
transaction can never succeed.

Given that, during my Tickets development, I changed the return code
from tefPAST_SEQ to temBAD_SEQUENCE.  Malformed transactions should
have 'tem' codes.

But, that change means the submitted code is transaction changing.
This commit returns us to the old behavior so, as far as I know, the
'Tx Change' label can be removed from the pull request.
```

scottschurr replied:
Yeah, my preference would be to return a `temBAD_SEQUENCE` here, as Nik noted.  However, historically, the zero value was not explicitly checked.  So the old behavior for any transaction seeing a Sequence of zero would be to return `tefPAST_SEQ`.  Vinnie asked me to adjust the pull request so it was not transaction changing.  Returning a `tefPAST_SEQ` here (and a couple of other places) was part of that effort.

You folks can get a sense of the nature of the change by looking at this commit: https://github.com/ripple/rippled/commit/4a898a91f881ea920b95d619a3d35d489118bcb1

Personally, I prefer returning `temBAD_SEQUENCE`.  A zero Sequence without a Ticket can never succeed.  But it is, literally, transaction changing.  But it's only transaction changing if someone messes up big time by giving a transaction a zero sequence and no Ticket.  You guys can decide on the right call to make.

nbougalis commented on src/ripple/app/tx/impl/CancelTicket.cpp:
```diff
@@ -41,49 +42,23 @@ class CancelTicket
-        return result;
+        // See if we can legitimately use the Ticket.  Note that canceling
+        // an expired Ticket is a tesSUCCESS.
+        return checkAndConsumeCancelTicket (mTxn, mTxnAccountID, mEngine);
```
`checkAndConsumeCancelTicket` is called only from here and nowhere else, so why not just inline the body in here? What problem is the indirection solving?

scottschurr replied:
I struggled a bit with the three circumstances where we consume tickets: a) replacement for Sequence in a transaction, b) explicit cancel of the ticket, and c) consume the Ticket for a `tec` error.  They felt like they had a lot of commonality.  But the specifics of each case gave each a unique twist.

Anyway, I ended up making three different free functions that share a number of helper functions and a helper struct.  All three of the free function implementations are in the same file so they can share the helpers and so the helpers are readily accessible to anyone that maintains the code.

I think it is helpful to keep `checkAndConsumeSeqTicket()` and `checkAndConsumeCancelTicket()` in the same file and adjacent to one another.  It makes it easier for a maintainer to see that there are two different situations for removing tickets, and makes it easier to see the differences and commonalities.  As Nik suggests in a different place, I'll probably get rid of the `consumeTicketForTecError()` free function and inline that the one place it is called.

If you folks have suggestions for better names for the free functions I'm all ears.  I'm not at all wedded to the names.  Perhaps `checkAndConsumeTicket()` and `checkAndCancelTicket()` are a bit better?

nbougalis commented on src/ripple/app/tx/impl/TransactionEngine.cpp:
```diff
@@ -133,12 +138,32 @@ TransactionEngine::applyTransaction (
+                            txn.getFieldAccount160 (sfAccount), this))
+                        {
+                            // Unable to consume Ticket.  Can't charge a 'tec'.
+                            terResult = tefNO_PERMISSION;
```
What do you mean "can't charge a tec"? A `tec` code means "claim a fee". If the comment means to suggest that a fee should be chanrged, then a `tec` is exactly what we need to return and returning a `tef` is wrong.

scottschurr replied:
I think you've made a good call here.  I'll try inlining `consumeTicketForTecError()` here.  You can see if you prefer the result.

vinniefalco commented on src/ripple/app/ledger/LedgerConsensus.h:
```diff
@@ -73,11 +74,28 @@ make_LedgerConsensus (
+void
+applyTransactions(std::shared_ptr<SHAMap> const& set, TransactionEngine& engine,
+                  Ledger::ref checkLedger,
+                  CanonicalTXSet& retriableTransactions, bool openLgr);
```
Suggested: `retriables`

vinniefalco commented on src/ripple/app/misc/CanonicalTXSet.cpp:
```diff
@@ -80,9 +62,22 @@ void CanonicalTXSet::push_back (STTx::ref txn)
+        ticketSeq = ticketID.getFieldU32 (sfSequence);
+    }
+
+    mMap.insert (
```
`emplace`?

vinniefalco commented on src/ripple/app/ledger/LedgerEntrySet.cpp:
```diff
@@ -2057,9 +2057,12 @@ adjustOwnerCount (LedgerEntrySet& view,
+
+    // The following tests rely on "current" and "adjusted" being unsigned.
+    static_assert (std::is_unsigned<decltype(current)>::value, "");
+    static_assert (std::is_unsigned<decltype(adjusted)>::value, "");
```
This looks robust

vinniefalco commented on src/ripple/app/ledger/tests/common_ledger.cpp:
```diff
@@ -432,29 +432,67 @@ trust(TestAccount& from, TestAccount const& issuer,
     newLCL->updateSkipList();
     newLCL->setClosed();
     newLCL->peekAccountStateMap()->flushDirty(
         hotACCOUNT_NODE, newLCL->getLedgerSeq());
```
Why `flushDirty`? This writes to the node store, do we want that?

scottschurr replied:
That's a fair question, and I don't know the answer.  `close_and_advance` has had a few hands in its creation.  The part you are referring to was present when I got here.  I've studied the function well enough to modify it, but not enough to truly understand it.

vinniefalco commented on src/ripple/app/tx/impl/LocalTxs.cpp:
```diff
@@ -61,48 +64,83 @@ class LocalTx
+        return txn_;
+    }
+
+    AccountID const& getAccountID () const
```
Just out of curiosity, and I'm not suggesting that we change this because in reality the `LocalTxs` class will go away, but what did we gain by making these accessor functions instead of just making the data members `public`?

scottschurr replied:
I think that would have been fine.  In this particular case I was just following the pre-existing pattern.  The minor advantage when getters are provided but not setters is that the interface behaves as though the contained data is `const` without it actually needing to be `const`.  I'm not saying that minor advantage justifies the complexity in this particular case.

vinniefalco commented on src/ripple/app/tx/impl/Transactor.cpp:
```diff
@@ -283,21 +335,24 @@ TER Transactor::apply ()
         mEngine->view().entryModify (mTxnAccount);
 
+    // We successfully passed the signature check.  Process any "tec" errors.
+    if (terHandler.hasTec()) return (terHandler.tec);
```
Two lines and why the parenthesis?

scottschurr commented:
@nbougalis and @vinniefalco: how do you feel about this pull request?  Should I squash it and rebase and then start pestering @JoelKatz for a code review?  Nik, you indicated you were going to spend more time looking over the code.  Have you done that?

The biggest question right now that I'm aware of is the "[FOLD] Maintain transaction backward compatibility:" commit (https://github.com/scottschurr/rippled/commit/4a898a91f881ea920b95d619a3d35d489118bcb1).  It arguably preserves transaction backward compatibility.  But it's just as easy to argue that the commit changes a more appropriate `tem` to a less appropriate `tef`.  A transaction with a zero sequence and no Ticket really is malformed and can never succeed in any ledger. If it were my call, I'd remove the [FOLD].  But I don't feel strongly about it.  You guys make the call.

scottschurr commented:
At Vinnie's suggestion I've modified the "[FOLD] Maintain transaction backward compatibility:" commit (scottschurr@4a898a9) to this:https://github.com/scottschurr/rippled/commit/5e44b6cae27fc62d3b3273cc64d8edca6a65eaa3.  The new version changes the return error code based on whether Tickets are disabled (backward compatibility) or enabled (which is a transaction change, so we can use the better code).

So at this point, as far as I know, I've addressed all concerns from Nik and Vinnie.

vinniefalco commented on src/ripple/app/tx/impl/CheckAndConsumeTicket.cpp:
```diff
@@ -0,0 +1,197 @@
+    SLE::ref sleOwner,
+    AccountID const& owner,
+    uint256 const& ticketIndex,
+    MetaView& view)
```
This should be `View&` instead. Use the abstract interfaces whenever possible (see the signature for `dirDelete`)

vinniefalco commented on src/ripple/app/misc/tests/CanonicalTXSet.test.cpp:
```diff
@@ -0,0 +1,158 @@
+
+        auto eqTest = [this] (Key const& lhs, Key const& rhs)
+        {
+            this->expect ((lhs <  rhs) == false);
```
Why does `CanonicalTXSet::Key` implement all the relational operators instead of just `<` ?

vinniefalco commented on src/ripple/app/misc/tests/CanonicalTXSet.test.cpp:
```diff
@@ -0,0 +1,158 @@
+    void test_txset_key_ordering()
+    {
+        // Verify that CanonicalTXSet::Key sorts as expected.
+        using Key = CanonicalTXSet::Key;
```
`CanonicalTXSet::Key` should be `private` since before this change no one is using it. And the key should not be exposed merely for testing.

scottschurr replied:
@vinniefalco: So your preference is to make `CanonicalTXSet::Key` private and remove the unit test?  I can do that, but I would prefer not to remove the test.  `CanonicalTXSet::Key` determines the order in which transactions are put in a final ledger.  To me, a test that verifies the `Key`'s behavior seems appropriate.  I certainly needed the test while I modified the key to support Tickets.

vinniefalco replied:
You could eliminate the key entirely:
```
// A StrictWeakOrdering for putting
// transactions in a canonical ordering
class CanonicalLess
{
private:
    ...
public:
    CanonicalLess (uint256 const& lastClosedLedgerDigest);
    template <class Object>
    bool operator()(std::shared_ptr<Object const> const& lhs,
        std::shared_ptr<Object const> const& rhs) const;
};
```

Now you can replace the member in CanonicalTxSet with `std::set<std::shared_ptr<STTx const>, CanonicalLess>` and still have access to the sort for your unit tests (for example, by using `std::sort` on a vector of transactions.

vinniefalco commented on src/ripple/app/misc/tests/CanonicalTXSet.test.cpp:
```diff
@@ -0,0 +1,158 @@
+        {
+            this->expect ((lhs <  rhs) == false);
+            this->expect ((lhs <= rhs) == true);
+            this->expect ((lhs >  rhs) == false);
```
Why `(lhs <= rhs) == true` instead of `lhs <= rhs`?

vinniefalco commented on src/ripple/app/ledger/tests/common_ledger.cpp:
```diff
@@ -435,28 +435,61 @@ trust(TestAccount& from, TestAccount const& issuer,
 void
 close_and_advance(Ledger::pointer& ledger, std::shared_ptr<Ledger const>& LCL)
```
*common_ledger* is deprecated. We should not add to it. Changes to existing tests or new tests should use *jtx*.

scottschurr replied:
It's true that _common_ledger_ is deprecated, but it's also providing the implementation of `jtx::advance` at the moment.  See https://github.com/ripple/rippled/blob/develop/src/ripple/test/jtx/impl/advance.cpp#L32.  From my perspective, until `close_and_advance` functionality is moved into `jtx`.  keeping all of `close_and_advance` together in one place will make it easier to move in the future.

vinniefalco replied:
`jtx::advance` should not be in jtx then, since it uses a deprecated interface

ximinez replied:
`jtx::advance` only calls into _common_ledger_ for simplicity and to avoid another copy/paste. If we want to make the push to deprecate _common_ledger_, we should *move* that code into `jtx::advance`. Additionally, the intention of `jtx::advance` was for specialized usages to create local overrides instead of proliferating a bunch of global overrides.

vinniefalco replied:
How do you create a local overide?

ximinez replied:
I meant "overload". So, eg, different parameters.

vinniefalco commented on src/ripple/app/ledger/tests/common_ledger.h:
```diff
@@ -236,11 +240,23 @@ trust(TestAccount& from, TestAccount const& issuer,
+// a `ter` or `tec`.
+void
+close_and_advance(Ledger::pointer& ledger, std::shared_ptr<Ledger const>& LCL,
+    std::uint32_t advSeconds, std::unique_ptr<LocalTxs>& retries);
```
Functionality for advancing the ledger and dealing with retries should be in jtx. And we should not use `LocalTxs` to store the list of retriables.

scottschurr replied:
1. I agree that functionality for advancing the ledger should be in `jtx`.  However, `Ledger_test.cpp` still relies on `common_ledger.cpp`.  Once `Ledger_test.cpp` is ported to `jtx`, then it should be possible to salvage the necessary parts out of `common_ledger` and move them to `jtx`.  To me it seems the time for that move is not yet right.
2. For the moment, I believe `LocalTxs` is the only appropriate choice to hold the list of retryables.  The mock relies on the `LocalTxs::sweep()` and `LocalTxs::apply()` interfaces to do its work.  Until those interfaces disappear, due to the Fees work, I believe this is the right choice.

vinniefalco commented on src/ripple/app/tx/impl/CheckAndConsumeTicket.cpp:
```diff
@@ -0,0 +1,197 @@
+
+namespace ripple {
+
+namespace
```
unnecessary

vinniefalco commented on src/ripple/app/tx/impl/CheckAndConsumeTicket.cpp:
```diff
@@ -0,0 +1,197 @@
+namespace
+{
+
+struct TicketContents
```
`using Contents = std::pair<AccountID, std::uint32>`

vinniefalco commented on src/ripple/app/tx/impl/CheckAndConsumeTicket.cpp:
```diff
@@ -0,0 +1,197 @@
+
+        if (txnEngine->getLedger()->getParentCloseTimeNC() >= expiration)
+            return true;
+    }
```
This conflicts significantly with changes being made to transaction processing. There's no visible `TransactionEngine`,  And no ledger object. The parent close time is on the BasicView. We might have to pass a `BasicView` instead.

scottschurr replied:
I can't make that change yet, since `GetParentCloseTimeNC()` is not yet available on `BasicView`.  Perhaps with the next rebase :-).

vinniefalco replied:
its really close :-) `BasicView::info().parentCloseTime`

vinniefalco commented on src/ripple/app/ledger/tests/common_ledger.cpp:
```diff
@@ -435,28 +435,61 @@ trust(TestAccount& from, TestAccount const& issuer,
+}
+
+void
+close_and_advance(Ledger::pointer& ledger, std::shared_ptr<Ledger const>& LCL,
```
I have concerns that we've duplicated functionality of `LedgerConsensusImp::accept` here for testing. The accept function is undergoing changes. And will have to reflect those changes here. It will probably break.

scottschurr replied:
@vinniefalco, I hear your concerns, but that is the nature of a mock for unit testing.  And in this case I'm making the mock more sophisticated so it can test problems that I found using my external python tests.  Performing these tests using only `npm` tests will be slow, since Tickets must expire.  By using this interface I can both programmatically advance time and simulate transaction retries (both of which I need for certain test cases).

vinniefalco commented on src/ripple/app/ledger/tests/common_ledger.cpp:
```diff
@@ -435,28 +435,61 @@ trust(TestAccount& from, TestAccount const& issuer,
+
+void
+close_and_advance(Ledger::pointer& ledger,
+    std::shared_ptr<Ledger const>& LCL, std::uint32_t advSeconds)
```
We don't need a whole function just to pass an empty container of transactions, just make callers do it.

scottschurr commented:
I'm closing this pull request.  There are enough unsolved problems with the Ticket design that I'm hesitant to introduce the Ticket-related complexity into the code base without clear solutions.  I'll keep a copy of this branch in case we decide to revisit Tickets in the future.
